gfx-examples/src/shadow.rs

```diff
@@ -190,32 +193,51 @@ impl framework::Example for Example {
         // Create the vertex and index buffers
         let vertex_size = mem::size_of::<Vertex>();
         let (cube_vertex_data, cube_index_data) = create_cube();
-        let cube_vertex_buf = Rc::new(device.create_buffer(&wgpu::BufferDescriptor {
-            size: (cube_vertex_data.len() * vertex_size) as u32,
-            usage: wgpu::BufferUsageFlags::VERTEX | wgpu::BufferUsageFlags::TRANSFER_DST,
-        }));
-        cube_vertex_buf.set_sub_data(0, framework::cast_slice(&cube_vertex_data));
-        let cube_index_buf = Rc::new(device.create_buffer(&wgpu::BufferDescriptor {
-            size: (cube_index_data.len() * 2) as u32,
-            usage: wgpu::BufferUsageFlags::INDEX | wgpu::BufferUsageFlags::TRANSFER_DST,
-        }));
-        cube_index_buf.set_sub_data(0, framework::cast_slice(&cube_index_data));
-
-        let (plane_vertex_data, plane_index_data) = create_plane(7);
-        let plane_vertex_buf = device.create_buffer(&wgpu::BufferDescriptor {
-            size: (plane_vertex_data.len() * vertex_size) as u32,
-            usage: wgpu::BufferUsageFlags::VERTEX | wgpu::BufferUsageFlags::TRANSFER_DST,
+        let cube_vertex_buf = Rc::new({
+            let (vertex_buf, data) = device.create_buffer_mapped(&wgpu::BufferDescriptor {
+                size: (cube_vertex_data.len() * vertex_size) as u32,
+                usage: wgpu::BufferUsageFlags::VERTEX | wgpu::BufferUsageFlags::TRANSFER_DST | wgpu::BufferUsageFlags::MAP_WRITE,
+            });
+            data.copy_from_slice(&cube_vertex_data);
+            vertex_buf.unmap();
```

Comment:
is there way to have `create_buffer_mapped` returning something like a guard that can be dereferenced into `&[T]` but also calls the related buffer `unmap()` in its own destructor? I think it should even be borrowing the buffer mutably for its lifetime.

Reply:
I don't believe there's a way to return both the buffer and a mutable borrow of the buffer at the same time (there probably is once Pin stabilises, but that seems like overkill).

I'm in favour of a MappedBufferGuard for map_read/write_async, but multiple return values makes things tricky for create_buffer_mapped

Reply:
ok, we don't have to borrow the buffer. We can return something that is not the buffer. The idea is to support a use case where the user fills up the contents right away, and then uses it for some recording, so there isn't any expectation that the recording happens at the same time as the buffer is mapped. I.e. this could be:
```rust
let vertex_buf = {
  let map = device.create_buffer_mapped(...);
  map.copy_from_slice(&some_data);
  map.finish() // turns it into `wgpu::Buffer`
};

Reply:
The separate `finish()` turns out to be painful, because copy_from_slice takes a mutable borrow on self. Take a look at what I've proposed in the new commit, and see what you think.

Reply:
I like what you added, and I think we can have the best of both worlds. We'd still want to expose the slice for any case that doesn't have a slice of data ready to go. For this to be done, we just need `impl DerefMut for CreateBufferMapped` that would deref it to `&mut [T]`. We can still have and use the convenience shortcut method `from_slice` that you added, while allowing the user to fill it up manually as well.

Reply:
I've added an explict `as_mut_slice()` accessor instead. Happy to change it to DerefMut if you prefer, but I liked the explicitness